### PR TITLE
[inductor] refactor: device dispatch inside do_bench

### PR DESCRIPTION
**Pull request by jgong5**
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #124021
* __->__ #125736

cc @voznesenskym @penguinwu @EikanWang @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

### CODE REVIEW

File: torch/_inductor/runtime/runtime_utils.py [hunk 1]
```diff
@@ -70,7 +70,18 @@ def get_max_y_grid():
...
+
+    args = list(fn_args)
+    args.extend(fn_kwargs.values())
+    if is_cpu_device(args):
+        return do_bench_cpu(lambda: fn(*fn_args, **fn_kwargs), **kwargs)
+    else:
+        return do_bench_gpu(lambda: fn(*fn_args, **fn_kwargs), **kwargs)
+
+
+def do_bench_gpu(fn, **kwargs):
```
Comment (shunting314) [inline torch/_inductor/runtime/runtime_utils.py:84]: Why change `*args` to fn? I think the former way is more flexible.

Comment (jgong5) [reply torch/_inductor/runtime/runtime_utils.py]: Fixed.

**Review by shunting314 — APPROVED** [review]

**Comment by izaitsevfb** [comment]
Hey @jgong5, please note that `TestBench` also needs to change:
https://github.com/pytorch/pytorch/blob/a174c536f8f32b41da9efa647e364196423468a5/test/inductor/test_inductor_utils.py#L24
```
in test_do_bench
    res = do_bench(self._bench_fn)
TypeError: do_bench() missing 2 required positional arguments: 'fn_args' and 'fn_kwargs'
```

**Comment by msaroufim** [comment]
This change also broke the ao CI https://github.com/pytorch/ao/actions/runs/9073922969/job/24931815576?pr=189#step:11:3465

**Comment by jgong5** [comment]
> Hey @jgong5, please note that `TestBench` also needs to change:

Thanks for the reminder. Let me fix it with a PR. But may I know why it was not failing the CI?

**Comment by jgong5** [comment]
> This change also broke the ao CI https://github.com/pytorch/ao/actions/runs/9073922969/job/24931815576?pr=189#step:11:3465

Apology for that. Is this what @jerryzh168 is fixing with https://github.com/pytorch/ao/pull/242?

**Comment by jerryzh168** [comment]
> > This change also broke the ao CI [pytorch/ao/actions/runs/9073922969/job/24931815576?pr=189#step:11:3465](https://github.com/pytorch/ao/actions/runs/9073922969/job/24931815576?pr=189#step:11:3465)
> 
> Apology for that. Is this what @jerryzh168 is fixing with [pytorch/ao#242](https://github.com/pytorch/ao/pull/242)?

yeah this PR should fix the issue for torchao, but going forward I feel maybe we want to have a more public do_bench API that everyone can rely on

**Comment by jgong5** [comment]
> > Hey @jgong5, please note that `TestBench` also needs to change:
> 
> Thanks for the reminder. Let me fix it with a PR. But may I know why it was not failing the CI?

Here is the PR to fix this: https://github.com/pytorch/pytorch/pull/126224. Please review. @izaitsevfb

**Pull request closed without merging**